nimsuggest/src/nimsuggest.nim

```diff
@@ -0,0 +1,269 @@
+#
+#
+#           The Nim Compiler
+#        (c) Copyright 2017 Andreas Rumpf
+#
+#    See the file "copying.txt", included in this
+#    distribution, for details about the copyright.
+#
+
+## Nimsuggest is a tool that helps to give editors IDE like capabilities.
+
+import strutils, os, parseopt, net, tables, times, communication, helpText, parsing
+
+import compiler/[renderer, options, commands, modules, passes, passaux, msgs, idents, modulegraphs,
+    lineinfos, cmdlinehelper, pathutils, condsyms]
+
+import globals, utils, emacs/emacs, execution, repl, types
+
+when not defined(nimcore):
+  {.error: "nimcore MUST be defined for Nim's core tooling".}
+
+# Do NOT import suggest. It will lead to weird bugs with
+# suggestionResultHook, because suggest.nim is included by sigmatch.
+# So we import that one instead.
+
+when defined(nimPreviewSlimSystem):
+  import std/typedthreads
+
+when defined(windows):
+  import winlean
+else:
+  import posix
+
+
+proc execCmd(cmdLineString: string; graph: ModuleGraph; cachedMsgs: CachedMsgs) =
+  let conf = graph.config
+
+  template sentinel() =
+    # send sentinel for the input reading thread:
+    results.send(Suggest(section: ideNone))
+
+  template toggle(sw) =
+    if sw in conf.globalOptions:
+      excl(conf.globalOptions, sw)
+    else:
+      incl(conf.globalOptions, sw)
+    sentinel()
+    return
+
+  template err() =
+    echo Help
+    sentinel()
+    return
+  let cmd = parseCommandLine(cmdLineString, string conf.projectFull)
+  ##If we can't parse an ideCmd then it's either a nimsuggest command or grabage
+  if cmd.ideCmd == ideNone:
+    case cmd.ideCmdString:
+    of "quit":
+      sentinel()
+      quit()
+    of "debug": toggle optIdeDebug
+    of "terse": toggle optIdeTerse
+    else: err()
```

Comment:
As you can see we re-parse the cmdString if the command isn't on the list the compiler knows about. 

This is where we use the toggle command which turns debug/ terse off. I believe this should likely trigger an early return. There isn't really any reason to try to execute the rest of the code if the command is a toggle.